ergo/distributions/point_density.py

```diff
@@ -0,0 +1,199 @@
+                init_np.cumsum(bin_probs), init_np.array([1.0])
+            )
+        self.scale = scale
+        self.normed_log_densities = np.log(self.normed_densities)
+
+    # Distribution
+
+    def pdf(self, x):
+        """
+        If x is out of distribution range, returns 0. Otherwise returns the
+        density at the lowest bin for which the lower bound of the bin
+        is greater than or equal to x.
+
+        :param x: The point in the distribution to get the density at
+        """
+        normed_x = self.scale.normalize_point(x)
+        bin = np.maximum(np.argmax(self.normed_xs >= normed_x), 0)
+        p = self.scale.denormalize_density(normed_x, self.normed_densities[bin])
+        return np.where((normed_x < 0) | (normed_x > 1), 0, p)
+
+    def logpdf(self, x):
+        return np.log(self.pdf(x))
+
+    def cdf(self, x):
+        normed_x = self.scale.normalize_point(x)
+        bin = np.maximum(np.argmax(self.normed_xs >= normed_x), 0)
+        c = self.cumulative_normed_ps[bin]
+        return np.where(normed_x < 0, 0, np.where(normed_x > 1, 1, c))
+
+    def ppf(self, q):
+        return self.scale.denormalize_point(
+            self.normed_xs[np.argmax(self.cumulative_normed_ps >= q)]
+        )
+
+    def sample(self):
+        raise NotImplementedError
+
+    # Scaled
+
+    def normalize(self):
+        return PointDensity(
+            self.normed_xs, self.normed_densities, scale=Scale(0, 1), normalized=True
+        )
+
+    def denormalize(self, scale: Scale):
+        return PointDensity(
+            self.normed_xs, self.normed_densities, scale=scale, normalized=True
+        )
+
+    # Structured
+
+    @classmethod
+    def structure(cls, params):
+        class_params, numeric_params = params
+        density_class, scale_classes = class_params
+        density_numeric, scale_numeric = numeric_params
+        return density_class(
+            xs=density_numeric[0],
+            densities=density_numeric[1],
+            scale=scale_classes[0].structure((scale_classes, scale_numeric)),
+            normalized=True,
+            traceable=True,
+            cumulative_normed_ps=density_numeric[2],
+        )
+
+    def destructure(self):
+        scale_classes, scale_numeric = self.scale.destructure()
+        class_params = (self.__class__, scale_classes)
+        self_numeric = self.normed_xs, self.normed_densities, self.cumulative_normed_ps
+        numeric_params = (self_numeric, scale_numeric)
+        return (class_params, numeric_params)
+
+    # Optimizable
+
+    @classmethod
+    def from_conditions(cls, *args, fixed_params=None, scale=None, **kwargs):
+        if scale is None:
+            # TODO: Should we do this?
+            scale = Scale(0, 1)
+        if fixed_params is None:
+            fixed_params = {
+                "xs": np.linspace(
+                    scale.low, scale.high, constants.point_density_default_num_points
+                )
+            }
```

Comment:
Should check where in the code base these can happen, address in the calling location, and consider requiring here.